openedx_learning/core/contents/models.py

```diff
@@ -95,92 +105,211 @@ def __str__(self):
-    # the character length, since UTF-8 encoding can use anywhere between 1-4
-    # bytes to represent any given character.
+    # This is the size of the file in bytes. This can be different than the
+    # character length of a text file, since UTF-8 encoding can use anywhere
+    # between 1-4 bytes to represent any given character.
     size = models.PositiveBigIntegerField(
         validators=[MaxValueValidator(MAX_FILE_SIZE)],
     )
 
-    # This should be manually set so that multiple RawContent rows being set in
-    # the same transaction are created with the same timestamp. The timestamp
-    # should be UTC.
-    created = manual_date_time_field()
+    # This hash value may be calculated using create_hash_digest from the
+    # openedx.lib.fields module. When storing text, we hash the UTF-8
+    # encoding of that text value, regardless of whether we also write it to a
+    # file or not.
+    hash_digest = hash_field()
 
-    # All content for the LearningPackage should be stored in files. See model
-    # docstring for more details on how to store this data in supplementary data
-    # models that offer better latency guarantees.
-    file = models.FileField(
+    # Do we have file data stored for this Content in our file storage backend?
+    has_file = models.BooleanField()
+
+    # The ``text`` field contains the text representation of the Content, if
+    # it is available. A blank value means means that we are storing text for
+    # this Content, and that text happens to be an empty string. A null value
+    # here means that we are not storing any text here, and the Content exists
+    # only in file form. It is an error for ``text`` to be None and ``has_file``
+    # to be False, since that would mean we haven't stored data anywhere at all.
+    #
+    # We annotate this because mypy doesn't recognize that ``text`` should be
+    # nullable when using MultiCollationTextField, but does the right thing for
+    # TextField. For more info, see:
+    #   https://github.com/openedx/openedx-learning/issues/152
+    text: models.TextField[str | None, str | None] = MultiCollationTextField(
+        blank=True,
         null=True,
-        storage=settings.OPENEDX_LEARNING.get("STORAGE", default_storage),  # type: ignore
+        max_length=MAX_TEXT_LENGTH,
+        # We don't really expect to ever sort by the text column, but we may
+        # want to do case-insensitive searches, so it's useful to have a case
+        # and accent insensitive collation.
+        db_collations={
+            "sqlite": "NOCASE",
+            "mysql": "utf8mb4_unicode_ci",
+        }
     )
 
+    # This should be manually set so that multiple Content rows being set in
+    # the same transaction are created with the same timestamp. The timestamp
+    # should be UTC.
+    created = manual_date_time_field()
+
     @cached_property
     def mime_type(self):
         return str(self.media_type)
```

Comment:
We should have a type hint here. Also, optionally, here's a docstring you could add.

```suggestion
    @cached_property
    """
    The IANA media type (a.k.a. MIME type) of the content stored in `text` and/or `file`.
    
    MIME types reference:
    https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
    """
    def mime_type(self) -> str:
        return str(self.media_type)
```